modules/core/src/test/java/org/apache/ignite/internal/processors/datastreamer/DataStreamerImplSelfTest.java

```diff
@@ -110,6 +115,246 @@
         return cfg;
     }
 
+    /**
+     * @throws Exception If failed.
+     */
+    public void testSimilarFuturePerBatch() throws Exception {
+        cnt = 0;
+
+        startGrids(2);
+
+        IgniteDataStreamer<Integer, Integer> dataLdr = grid(0).dataStreamer(DEFAULT_CACHE_NAME);
+
+        IgniteFuture curFut;
+
+        IgniteFuture prevFut = dataLdr.addData(1, 1);
+
+        for (int i = 2; i <= DATA_AMOUNT; i++) {
+            curFut = dataLdr.addData(i, i);
+
+            if (i % IgniteDataStreamer.DFLT_PER_THREAD_BUFFER_SIZE == 1)
+                assertFalse(curFut == prevFut);
+            else
+                assertTrue(curFut == prevFut);
+
+            prevFut = curFut;
+        }
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testFuturesAmountMultiThread() throws Exception {
+        cnt = 0;
+
+        startGrids(2);
+
+        IgniteDataStreamer<Integer, Integer> dataLdr = grid(0).dataStreamer(DEFAULT_CACHE_NAME);
+
+        int threadsNum = 3;
+
+        final CountDownLatch l1 = new CountDownLatch(threadsNum);
+
+        HashSet<IgniteFuture> uniFut = new HashSet<>();
+
+        multithreadedAsync(new Callable<Object>() {
+            @Override public Object call() throws Exception {
+                for (int i = 1; i <= DATA_AMOUNT; i++)
+                    uniFut.add(dataLdr.addData(i, i));
+
+                l1.countDown();
+
+                return null;
+            }
+        }, threadsNum);
+
+        l1.await();
+
+        final double futAmount = Math.ceil((double) DATA_AMOUNT /
+                IgniteDataStreamer.DFLT_PER_THREAD_BUFFER_SIZE) * threadsNum;
+
+        assertTrue(uniFut.size() == futAmount);
+    }
+
+    /**
+     *@throws Exception If failed.
+     */
+    public void testStreamingAfterFillingBatch() throws Exception {
+        cnt = 0;
```

Comment:
Why cnt is a class field?

Reply:
this is counter of started grid. After each start this counter increments

Reply:
Why we have to modify it in each test? This looks like a misusage to me.

Reply:
Because foundation of this class is not mine.